Improve `from` behavior for mandatory fields on partial Modifiable instances

stevenschlansker opened the pull request:
Currently the Modifiable implementations check if a field has been set and throw if not, which prevents a partial Modifiable object seeding a new builder.

This PR adds a Builder and Modifiable `from` overload for the Modifiable type, as well as a dynamic `instanceof` check since we can't rely on people statically calling the overload.  The overload then checks `isSet` on mandatory attributes before attempting to copy them, avoiding the exception.

elucash commented:
Thank you @stevenschlansker for your work! I think that having that kind of code should be quite a good contribution. Let us know when you will have something you're comfortable to review.
As a hint, if you need to share some templates between `Immutables.generator` and `Modifiables.generator`, you can declare them in `Immutables.generator` as `[template public name.....]` then add corresponding signature in `Immutables.java` as `abstract Templates.Invokable name....();` and this will enable to use it as `[im.name..` in `Modifiables.generator` as it's already "imported" there using `im.` prefix

stevenschlansker commented:
Thanks @elucash for the hint, I was able to factor out the dynamic check.  I wanted to factor out the `if(isset) { buildFromAttribute }` block, but unfortunately `buildFromAttribute` is defined differently in Immutables vs Modifiables, so factoring that out of Modifiables causes it to use the wrong method names from Immutables.  Is it allowed to pass a template (function) as data?  i.e.
```
[template public doSomething Invokable attributeBuilder Attribute v]
[attributeBuilder v]
[/template]

[doSomething buildFromAttribute v]
```
If not, doing further factoring of this PR may require reorganizing existing code, which could be a followup improvement :)

elucash commented:
@stevenschlansker it is possible to pass fragments/templates to other templates using Invokable as a parameter, here's example https://github.com/immutables/immutables/pull/843#issuecomment-428390972

stevenschlansker commented:
This template engine is arcane indeed!  I will give that an attempt.

stevenschlansker commented:
That wasn't too bad!  @elucash I think this version is ready for review.

elucash commented on value-processor/src/org/immutables/value/processor/Modifiables.generator:
```diff
@@ -401,11 +401,31 @@ private void from(Object object) {
+ */
+public [thisReturnType type] [type.names.from]([type.typeModifiable] instance) {
+  [im.requireNonNull type](instance, "instance");
+  final [type.typeModifiable] cast = ([type.typeModifiable]) instance;
```
don't need to cast here, param already modifiable

elucash commented on value-processor/src/org/immutables/value/processor/Immutables.generator:
```diff
@@ -1307,11 +1309,27 @@ public interface BuildFinal[type.generics] {
   }
+  [/if]
+  [if type.kind.isModifiable]
+  /**
```
probably need to add a blank line inside `[if` before the method. For methods, nested types and sometimes group of fields, when we have conditional output we put separating blank lines inside `[if` or `[for` directives, but before the methods so generated code would have needed blank lines and will not contain extra blank lines if optional code generated and when not.

elucash commented on value-processor/src/org/immutables/value/processor/Immutables.generator:
```diff
@@ -1253,6 +1253,7 @@ public interface BuildFinal[type.generics] {
   @SuppressWarnings("unchecked")
 [/if]
   private void from(Object object) {
+    [dynamicFromModifiableCheck type 'object' '']
```
`return ;` with extra space would be a little ugly here, maybe add some conditions inside `dynamicFromModifiableCheck`, like `return[if returnValue] [returnValue][/if];`

elucash commented on value-processor/src/org/immutables/value/processor/Immutables.generator:
```diff
@@ -2479,6 +2497,25 @@ this.[n] = [valueFromValue v][invokeSuper v].[v.names.get]()[/valueFromValue];
+  [buildAttribute v]
+}
+[else]
+  [buildAttribute v]
```
is this extra indentation, please, check generated source so it have proper formatting for both with `if ({`  and without

elucash commented on value-processor/src/org/immutables/value/processor/Immutables.generator:
```diff
@@ -2479,6 +2497,25 @@ this.[n] = [valueFromValue v][invokeSuper v].[v.names.get]()[/valueFromValue];
+
+[template public dynamicFromModifiableCheck Type type String varName String returnValue]
+[if type.kind.isModifiable]
+if ([varName] instanceof [type.typeModifiable.relativeRaw]) {
```
maybe as a matter of style/our preference we should add `[type.generics.unknown]` after `.relativeRaw]` so if type have generic parameters, it will output trailing `<?>` after raw type name (we try to avoid any use or raw types, even in instanceof I prefer to check it agains generic type with unknown arguments).
This also prompts to have at least one compilation test for value type having generic type parameters. We need check that we don't have new unsuppressed unchecked cast warnings in generated code.

elucash commented:
Thank you @stevenschlansker, really good work! there are some minor review comments to check, but overall we're good to go

stevenschlansker commented:
@elucash thanks for the review, I believe I've addressed all of your comments.  I also checked the indentation on the generated builder and it seems OK.

elucash commented:
Looks really cool! Thank you @stevenschlansker ! merging
